superset-frontend/src/dashboard/containers/DashboardPage.tsx

```diff
@@ -153,16 +154,47 @@ const DashboardPage: FC = () => {
   }, [readyToRender]);
 
   useEffect(() => {
-    if (readyToRender) {
-      if (!isDashboardHydrated.current) {
-        isDashboardHydrated.current = true;
-        if (isFeatureEnabled(FeatureFlag.DASHBOARD_NATIVE_FILTERS_SET)) {
-          // only initialize filterset once
-          dispatch(getFilterSets(id));
+    // eslint-disable-next-line consistent-return
+    async function getDataMaskApplied() {
+      const nativeFilterKeyValue = getUrlParam(URL_PARAMS.nativeFiltersKey);
+      let dataMaskFromUrl = nativeFilterKeyValue || {};
+
+      const isOldRison = getUrlParam(URL_PARAMS.nativeFilters);
+      // check if key from key_value api and get datamask
+      if (nativeFilterKeyValue) {
+        try {
+          dataMaskFromUrl = await getFilterValue(
+            dashboard?.id,
+            nativeFilterKeyValue,
+          );
+        } catch (err) {
+          return null;
         }
+        // else get old old rison string and set as datamask
+      } else if (isOldRison && dashboard?.id) {
+        dataMaskFromUrl = isOldRison;
+      }
+
+      if (readyToRender) {
+        if (!isDashboardHydrated.current) {
+          isDashboardHydrated.current = true;
+          if (isFeatureEnabled(FeatureFlag.DASHBOARD_NATIVE_FILTERS_SET)) {
+            // only initialize filterset once
+            dispatch(getFilterSets(id));
+          }
+        }
+        dispatch(
+          hydrateDashboard(
+            dashboard,
+            charts,
+            filterboxMigrationState,
+            dataMaskFromUrl,
+          ),
+        );
       }
-      dispatch(hydrateDashboard(dashboard, charts, filterboxMigrationState));
+      return null;
     }
+    if (dashboard?.id) getDataMaskApplied();
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [readyToRender, filterboxMigrationState]);
```

Comment:
a little code style improvement. 
1. We don't have to catch exceptions here(catch exception logic in the API layer) 
2. We can make a utility function, e.g.:`isCachedFilterKey`, `isFilterObject`
3. Dashboard id default value is `0`. If the dashboard has an id greater than 0, it will be a validated id.

```suggestion
  useEffect(() => {
    async function getDataMaskApplied(obj) {
      await getFilterValue(id, obj);
    }

    let dataMaskFromUrl;
    if (id > 0 && getUrlParam(URL_PARAMS.nativeFilters)) {
      dataMaskFromUrl = getUrlParam(URL_PARAMS.nativeFilters);
    }
    if (id > 0 && getUrlParam(URL_PARAMS.nativeFiltersKey)) {
      dataMaskFromUrl = getDataMaskApplied(getUrlParam(URL_PARAMS.nativeFiltersKey));
    }
    if (readyToRender) {
      if (!isDashboardHydrated.current) {
        isDashboardHydrated.current = true;
        if (isFeatureEnabled(FeatureFlag.DASHBOARD_NATIVE_FILTERS_SET)) {
          // only initialize filterset once
          dispatch(getFilterSets(id));
        }
      }
      dispatch(
        hydrateDashboard(
          dashboard,
          charts,
          filterboxMigrationState,
          dataMaskFromUrl,
        ),
      );
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [readyToRender, filterboxMigrationState]);
```

Reply:
This block still needs to be wrapped in async function other wise we end up with race conditions where it doesn't get the prev datamask and return empty object.